source/simulator/helper_functions.cc

```diff
@@ -1728,6 +1728,107 @@ namespace aspect
+    // interface between cells), as we loop over all cells, and then over all degrees of freedom
+    // on each cell. Although this means we do some additional work, the results are still
+    // correct, as we never read from distributed_vector inside the loop over all cells.
+    // We initialize the material model inputs object using the solution vector
+    // on every cell, compute the update, and then on every cell put the result into the
+    // distributed_vector vector. Only after the loop over all cells do we copy distributed_vector
+    // back onto the solution vector.
+    // So even though we touch some DoF twice, we always start from the same value, compute the
+    // same value, and then overwrite the same value in distributed_vector.
+    // TODO: make this more effective
+    typename DoFHandler<dim>::active_cell_iterator cell = dof_handler.begin_active(),
+                                                   endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit (cell);
+          cell->get_dof_indices (local_dof_indices);
+          in.reinit(fe_values, cell, introspection, solution);
+
+          // Make the reaction time steps: We have to update the values of compositional fields and the temperature.
+          // Because temperature and composition might use different finite elements, we loop through their elements
+          // separately, and update the temperature and the compositions for both.
+          // We can reuse the same material model inputs and outputs structure for each reaction time step.
+          // We store the computed updates to temperature and composition in a separate (accumulated_reactions) vector,
+          // so that we can later copy it over to the solution vector.
+
+          // Loop over composition element
+          material_model->evaluate(in, out);
+
+          // copy material properties into the compositional fields
+          for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+            for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
+              {
+                const unsigned int composition_idx
+                  = dof_handler.get_fe().component_to_system_index(introspection.component_indices.compositional_fields[c],
+                                                                   /*dof index within component=*/ j);
+
+                // skip entries that are not locally owned:
+                if (dof_handler.locally_owned_dofs().is_element(local_dof_indices[composition_idx]))
+                  {
+                    distributed_vector(local_dof_indices[composition_idx]) = copy_outputs->copy_properties[j][c];;
+                  }
+              }
+        }
+
+    // put the final values into the solution vector
+    for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
```

Comment:
This is the place I was talking about above. All the fields are copied, and not just the ones you want. So here and in line 1804, you want to remove the loop (and c will just be a function argument).